Wells 3d 1d

alessiofumagalli opened the pull request:
Add some features to couple 1d wells with the 3d rock matrix. An ADTree is introduced for fast geometrical searching.

IvarStefansson commented on src/porepy/utils/adtree.py:
```diff
@@ -0,0 +1,438 @@
+        node_id = 0
+        # List of possible intersections
+        found: List[int] = []
+        # List of right parts of the tree that need to be checked
```
```suggestion
        # List of right parts of the tree that need to be checked.
```

IvarStefansson commented on src/porepy/utils/adtree.py:
```diff
@@ -0,0 +1,438 @@
+        # List of possible intersections
+        found: List[int] = []
+        # List of right parts of the tree that need to be checked
+        # each element contains the node_id (position in self.nodes)
```
```suggestion
        # Each element contains the node_id (position in self.nodes)
```

IvarStefansson commented on src/porepy/utils/adtree.py:
```diff
@@ -0,0 +1,438 @@
+        return np.empty(0)
+
+    def from_grid(self, g: pp.Grid, only_cells: Optional[np.ndarray] = None) -> None:
+        """Function that construct the tree from a grid by adding one cells at the time.
```
```suggestion
        """Function that constructs the tree from a grid by adding one cell at a time.
```

IvarStefansson commented on src/porepy/utils/adtree.py:
```diff
@@ -0,0 +1,438 @@
+        """Function that construct the tree from a grid by adding one cells at the time.
+
+        If a portion of the cells should be considered due to some a-priori estimates of the
+        searching elements.
```
Isn't this a description of the only_cells parameter?

alessiofumagalli replied:
correct, I'll fix it

alessiofumagalli commented on src/porepy/fracs/wells_3d.py:
```diff
@@ -436,6 +437,110 @@ def compute_well_fracture_intersections(
+    tree.from_grid(g_max, cells)
+
+    # Extract the grids of the wells of co-dimension 2
+    gs_w = gb.grids_of_dimension(dim_max - 2)
```
do you think I can use the fact that a specific grid of dimension 1 has the attribute well_num?

alessiofumagalli replied:
okay, I used the fact that a well has the attribute well_num to filter

alessiofumagalli commented on src/porepy/utils/adtree.py:
```diff
@@ -0,0 +1,438 @@
+        key (Any): any key related to the node
+        box (np.ndarray): the bounding box associated to the node
+        child (list): list of identification of right and left children, if a children is not
+            present is marked as -1. It's not the key but the position of the children in the
```
Okay, I've removed it

alessiofumagalli commented:
@keileg and @IvarStefansson , thanks for your help to improve the pr. I still have some issues to uniform the setting of the mortar grid without the introduction of ad-hoc functions. In fact, MortarGrid._init_projections assumes that the grids are matching which is not my case. I think when @IvarStefansson implemented the well-fracture coupling only one dof of the well was involved and it simplified some of the aspects. What do you suggest? Adding an extra parameter to the mortar class, like primary_matching = True and handle at line 793 in a different way the case of non-matching primary grid.

keileg commented:
> I still have some issues to uniform the setting of the mortar grid without the introduction of ad-hoc functions. In fact, MortarGrid._init_projections assumes that the grids are matching which is not my case. I think when @IvarStefansson implemented the well-fracture coupling only one dof of the well was involved and it simplified some of the aspects. 

I still don't see the problem. You have a mapping from cells in the 3d grid to cells in the 1d grid, is that not correct? So if you make this the primary_secondary_map, will that not work? There should be no assumptions in the mortar grid that the dofs are cells and faces - enforcing this depends on how the projections are used when setting up equations.

alessiofumagalli commented on src/porepy/geometry/intersections.py:
```diff
@@ -1338,6 +1338,142 @@ def vector_pointset_point(a, b, tol=1e-8):
+            are both inside the polyhedron
+        2 - the segment has only one vertex in the polyhedron
+        3 - the segment is completely outside the polyhedron
+        4 - the segment has in intersection but both vertices are outside the polyhedron.
```
I've add the extra assumptions in the documentation.

keileg commented:
I will have a look later. TBC.

The tests should go into the same file. If the unclarity is caused by pytest instead of unittest, just try it - I like it much better.

alessiofumagalli commented on src/porepy/fracs/wells_3d.py:
```diff
@@ -436,6 +440,110 @@ def compute_well_fracture_intersections(
+
+        # create the mortar grid and set the maps
+        side_g = {pp.grids.mortar_grid.MortarSides.LEFT_SIDE: g_w.copy()}
+        mg = pp.MortarGrid(g_w.dim, side_g, codim=g_max.dim - g_w.dim)
```
okay, I've changed removing the new function in the MortarGrid

alessiofumagalli replied:
if I give the primary_secondary_map then the constructor calls the method _init_projections which creates problems since it assumes a matching coupling between the primary and secondary grids. in this way I don't have this problem

alessiofumagalli commented:
okay, I finish to answer to your doubts and requests. Let me know what do you think. Thanks again.

IvarStefansson reviewed (commented):
I had a quick second look. Am I right that there are noe specific questions for me?

alessiofumagalli commented:
@IvarStefansson, not really only if you can double check the changes I've done after your comments. Thanks :)

IvarStefansson reviewed (commented):
The changes seem good to me

keileg reviewed (commented):
I'm mostly happy. My remaining comments still relate to mortar grids, I think it should be sufficient to use only init (see detailed comments).

One question: Have you removed the suggested changes in the MortarGrid class?

alessiofumagalli commented:
@keileg , I've added the well_num to the grid and I noticed that these parameters specific for some cases are not copied with the method Grid.copy(). Should we include also them?
The additions I did to the MortarGrid are now removed, I'll answer to your other points in the next few minutes.

alessiofumagalli commented on src/porepy/fracs/wells_3d.py:
```diff
@@ -561,7 +675,7 @@ def _add_fracture_2_intersection_edge(
         (np.ones(1, dtype=bool), (cell_l, cell_h)), shape=(g_l.num_cells, g_h.num_cells)
     )
-    _add_edge(g_l, g_h, gb, cell_cell_map)
+    _add_edge(0, g_l, g_h, gb, cell_cell_map)
```
correct, the error is in the type of the function _add_edge

alessiofumagalli commented on src/porepy/fracs/wells_3d.py:
```diff
@@ -594,6 +712,7 @@ def _add_edge(
     side_g = {pp.grids.mortar_grid.MortarSides.LEFT_SIDE: g_l.copy()}
-    mg = pp.MortarGrid(0, side_g, primary_secondary_map, codim=codim)
+    mg = pp.MortarGrid(dim, side_g, primary_secondary_map, codim=codim)
+    mg._primary_to_mortar_int = primary_secondary_map
```
I don't use this function anymore in the compute_well_rock_matrix_intersections method, but I fill the parameters I need more explicitly.

alessiofumagalli commented:
The only part I cannot really come to your requests is related to the usage of the MortarGrid.__init__, the problem is related to the fact that if the map between the primary and secondary grid is provided it calls the method MortarGrid._init_projections. Specifically, line 770 creates the problem.

keileg commented:
@alessiofumagalli You mean the assertion that the mortar grid and the secondary grid should have the same number of cells? So, how have you defined the mortar grid in this case?

alessiofumagalli commented:
the mortar grid is the same as the secondary grid, the primary grid is non-matching with the mortar grid. that's the main issue when I tried to use the constructor of the MortarGrid

keileg commented:
Okay, let's leave it as it is, and possibly revisit at some later point. For the remaining mypy error, I think that a simple #type: ignore is okay here.

alessiofumagalli commented:
okay, thanks. I fix the mypy error now.

keileg commented on src/porepy/grids/mortar_grid.py:
```diff
@@ -675,6 +675,32 @@ def sign_of_mortar_sides(self, nd: int = 1) -> sps.spmatrix:
+    ) -> None:
+        """Set the int projection matrices from primary and secondary to the mortar grid.
+        Useful when the constructor is called with primary_secondary as None, also useful
+        when the grids are not matching to each other.
```
Let's leave this as it is now.

keileg reviewed (approved):
We'll merge this once the final mypy issue is resolved.
